Tools/jit/build.py

```diff
@@ -0,0 +1,699 @@
+        target = MachO("x86_64-apple-darwin", prefix="_")
+    elif re.fullmatch(r"x86_64-pc-windows-msvc", host):
+        target = COFF("x86_64-pc-windows-msvc")
+    elif re.fullmatch(r"x86_64-.*-linux-gnu", host):
+        target = ELF("x86_64-unknown-linux-gnu")
+    else:
+        raise ValueError(host)
+    return dataclasses.replace(target, debug=debug, verbose=verbose)
+
+
+def dump_header() -> typing.Iterator[str]:
+    yield f"// $ {shlex.join([sys.executable, *sys.argv])}"
+    yield ""
+    yield "typedef enum {"
+    for kind in sorted(typing.get_args(schema.HoleKind)):
+        yield f"    HoleKind_{kind},"
+    yield "} HoleKind;"
+    yield ""
+    yield "typedef enum {"
+    for value in sorted(HoleValue, key=lambda value: value.name):
+        yield f"    HoleValue_{value.name},"
+    yield "} HoleValue;"
+    yield ""
+    yield "typedef struct {"
+    yield "    const uint64_t offset;"
+    yield "    const HoleKind kind;"
+    yield "    const HoleValue value;"
+    yield "    const void *symbol;"
+    yield "    const uint64_t addend;"
+    yield "} Hole;"
+    yield ""
+    yield "typedef struct {"
+    yield "    const size_t body_size;"
+    yield "    const unsigned char * const body;"
+    yield "    const size_t holes_size;"
+    yield "    const Hole * const holes;"
+    yield "} Stencil;"
+    yield ""
+    yield "typedef struct {"
+    yield "    const Stencil code;"
+    yield "    const Stencil data;"
+    yield "} StencilGroup;"
+    yield ""
+
+
+def dump_footer(opnames: list[str]) -> typing.Iterator[str]:
+    yield "#define INIT_STENCIL(STENCIL) {                         \\"
+    yield "    .body_size = Py_ARRAY_LENGTH(STENCIL##_body) - 1,   \\"
+    yield "    .body = STENCIL##_body,                             \\"
+    yield "    .holes_size = Py_ARRAY_LENGTH(STENCIL##_holes) - 1, \\"
+    yield "    .holes = STENCIL##_holes,                           \\"
+    yield "}"
+    yield ""
+    yield "#define INIT_STENCIL_GROUP(OP) {     \\"
+    yield "    .code = INIT_STENCIL(OP##_code), \\"
+    yield "    .data = INIT_STENCIL(OP##_data), \\"
+    yield "}"
+    yield ""
+    yield "static const StencilGroup stencil_groups[512] = {"
+    for opname in opnames:
+        yield f"    [{opname}] = INIT_STENCIL_GROUP({opname}),"
+    yield "};"
+    yield ""
+    yield "#define GET_PATCHES() { \\"
+    for value in sorted(HoleValue, key=lambda value: value.name):
+        yield f"    [HoleValue_{value.name}] = (uint64_t)0xBADBADBADBADBADB, \\"
+    yield "}"
+
+
+def dump(stencil_groups: dict[str, StencilGroup]) -> typing.Iterator[str]:
+    yield from dump_header()
+    opnames = []
+    for opname, stencil in sorted(stencil_groups.items()):
+        opnames.append(opname)
+        yield f"// {opname}"
+        assert stencil.code
+        for line in stencil.code.disassembly:
+            yield f"// {line}"
+        body = ", ".join(f"0x{byte:02x}" for byte in stencil.code.body)
```

Comment:
This is probably the most beside-the-point comment ever, but we may want to line wrap these long arrays when we output them.  It doesn't have to be at 80 characters, but something like 160 characters maybe.  Emacs slows wayyyyy down with very long lines -- something about how the syntax highlighter works, I guess.

Reply:
Sure! `jit_stencils.h` is now 45k lines on debug builds, though... hope you like scrolling! ;)

Reply:
I'd much rather scroll vertically than horizontally. :-)

Other large generated files include parser.c (42k LoC) and deepfreeze.c (120k).